expression/src/functions.rs

```diff
@@ -0,0 +1,352 @@
+use crate::{
+    codegen::{DataType, Identifier},
+    error::ExpressionSemanticError,
+};
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use std::{collections::HashMap, hash::Hash, sync::OnceLock};
+
+type Result<T, E = ExpressionSemanticError> = std::result::Result<T, E>;
+
+/// A function generator that can be used to generate a function
+/// that can be used in an expression.
+///
+/// It checks whether the input arguments are valid.
+///
+pub struct FunctionGenerator {
+    /// User-facing name of the function
+    name: &'static str,
+
+    /// Method to validate the input arguments.
+    /// Return true if the arguments are valid.
+    generate_fn: fn(&str, &[DataType]) -> Result<Function>,
+}
+
+impl FunctionGenerator {
+    pub fn generate(&self, args: &[DataType]) -> Result<Function> {
+        (self.generate_fn)(self.name, args)
+    }
+}
+
+#[derive(Debug, Clone, Eq)]
+pub struct Function {
+    /// The call name of the function in the expression src code
+    name: Identifier,
+
+    /// Function signature
+    signature: Vec<DataType>,
+
+    /// Defines the output [`DataType`] of this function
+    output_type: DataType,
+
+    /// Write the function to a token stream
+    token_fn: fn(&Self, &mut TokenStream) -> (),
+}
+
+impl Function {
+    pub fn name(&self) -> &Identifier {
+        &self.name
+    }
+
+    pub fn output_type(&self) -> DataType {
+        self.output_type
+    }
+}
+
+impl ToTokens for Function {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        (self.token_fn)(self, tokens);
+    }
+}
+
+/// We generate a unique name, so we can compare [`Function`]s by name.
+mod cmp_by_name {
+    use super::*;
+
+    impl PartialEq for Function {
+        fn eq(&self, other: &Self) -> bool {
+            self.name == other.name
+        }
+    }
+
+    impl Ord for Function {
+        fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+            self.name.cmp(&other.name)
+        }
+    }
+
+    impl PartialOrd for Function {
+        fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+            Some(self.name.cmp(&other.name))
+        }
+    }
+
+    impl Hash for Function {
+        fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+            self.name.hash(state);
+        }
+    }
+}
+
+pub static FUNCTIONS: OnceLock<HashMap<&'static str, FunctionGenerator>> = OnceLock::new();
+
+/// Add a function generator for a function that returns a [`DataType::Number`] constant.
```

Comment:
is this a generator or does it add a generator to the hashmap?

Reply:
It is a generator for a function definition which itself is a generator for function source code :smile: